## Step 1: Set up a Jupyter Notebook
The aim of this section is to connect PostgreSQL with your Jupyter Notebook by leveraging Python Pandas. As mentioned earlier, you will need specific Python libraries to connect your Jupyter Notebook to PostgreSQL. 

You need to set up a conda environment (a directory that holds a collection of conda packages that you installed) on your computer with the Anaconda terminal. This will create the connection between PostgreSQL and Jupyter Notebook. 

Watch this video and follow along to learn the steps involved in creating a conda environment with Windows 10, but the instructions are similar for a Mac or Linux operating system. You can pause the video if needed to type in code snippets or go back to a previous timestamp if something is unclear. As long as, in the end, you understand and successfully set up the conda environment. In this video, you will learn how to:

- set up the conda environment
- install the necessary libraries into the conda environment
- activate the conda environment. 

Preparing the Conda environment is essential to have a smooth connection between PostgreSQL and your Jupyter Notebook. 

In case you are using Windows, navigate to the start menu and select the Anaconda terminal, also known as Anaconda Prompt. Type the following into the terminal;
- conda create -n sql python
- conda activate sql
- conda install ipython-sql sqlalchemy psycopg2 notebook pandas -c conda-forge

## Step 2: Import the Chinook database
First, you have to inform the Jupyter Notebook that you will use SQL for this session. The ipython-sql library helps to load packages into the Jupyter Notebook and will not work in any other Python environment or IDE. It is specifically designed for Jupyter Notebook. The code snippet can be written as:

- Indicate to Python to utilise SQL with the %load_ext sql command. The % is a Jupyter Notebook predefined function which allows us to replace traditional code snippets with more concise ones. It is also known as a magic command.
- Import the operating system package (os).
- Specify your PostgreSQL login details as:
  - host='localhost'
  - database='chinook'
  - user='postgres'
  -password= the password you set when you installed PostgreSQL.
- Create a variable (connection_string) to store the connection request. Note that the connection string is the path to connect to an existing database with the specified password, username and hostname.
- Specify the magic command (%sql) and ask for connection status with $connection_string. The output will indicate the connection status and the name of the database it has been connected to.
- Execute the code.

In [1]:
# The % is a magic command in Jupyter Notebook, it helps to load packages.
# It will not work in a regular Python environment, only in Jupyter Notebook.
# When the % is used in front, it indicates the use of SQL.
%load_ext sql

# Import operating system and the Anaconda environment.
import os

# PostgreSQL credentials.
host = 'localhost'
database = 'Chinook'
user = 'postgres'
# Your own password.
password = 'Pass123'     # Specify your own password.

# Test connection between PostgreSQL and Jupyter Notebook.
connection_string = f"postgresql://{user}:{password}@{host}/{database}"

# Determine connection status.
%sql $connection_string  



'Connected: postgres@Chinook'

In [2]:
# clear old tables if rerunning all code
%sql DROP TABLE artist_genre, genre_media, most_popular_artists, most_popular_artists

 * postgresql://postgres:***@localhost/Chinook
(psycopg2.errors.UndefinedTable) table "artist_genre" does not exist

[SQL: DROP TABLE artist_genre, genre_media, most_popular_artists, most_popular_artists]
(Background on this error at: https://sqlalche.me/e/14/f405)


### Confirm connection
We've set up the conda environment and successfully connected it to the Chinook database. Let’s query the database to establish the connection and sense-check the database.

Follow these prompts:

- Indicate to Jupyter Notebook that you'll use SQL syntax with the %%sql command. Take note, there are two kinds of magic commands:
- Line magic commands (start with %). This is where inputs are provided following the command in the same line. This is also a Modulo Operator that returns the remainder when dividing a left-hand operand by the right-hand operand.
- Cell magic commands (start with %%). This is where the content in the entire cell becomes the input.
- Specify the query with SQL syntax and the SELECT statement. Remember that if the column name includes a capital letter, you need to have double quotation marks ("") with SQL.
- Query all the values FROM the Artist table, but LIMIT the query to 10 rows for illustrative purposes.

In [3]:
# The % is a magic command in Jupyter Notebook, it helps to load packages.
%sql SELECT * from "Album" LIMIT 10


 * postgresql://postgres:***@localhost/Chinook
10 rows affected.


[(1, 'For Those About To Rock We Salute You', 1),
 (2, 'Balls to the Wall', 2),
 (3, 'Restless and Wild', 2),
 (4, 'Let There Be Rock', 1),
 (5, 'Big Ones', 3),
 (6, 'Jagged Little Pill', 4),
 (7, 'Facelift', 5),
 (8, 'Warner 25 Anos', 6),
 (9, 'Plays Metallica By Four Cellos', 7),
 (10, 'Audioslave', 8)]

In [4]:
# The % is a magic command in Jupyter Notebook, it helps to load packages.
%sql SELECT * from "Customer" Limit 10

 * postgresql://postgres:***@localhost/Chinook
10 rows affected.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4),
 (5, 'Franti\x9aek', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4),
 (6, 'Helena', 'Holý', None, 'Rilská 3174/6', 'Prague', None, 'Czech Republic', '14300', '+420 2 4177 0449', None, 'hholy@gmail.com', 5),
 (7, 'Astrid', 'Gruber', None, 'Rotenturmstraße 4, 1010 Innere Stadt', 'Vienne', None, 'Austria', '1010', '+43 01 5134505', None, 'astrid.gruber@apple.at', 5),
 (8, 'Daan', 'Peeters', None, 'Grétrystraat 63', 'Brussels', None, 'Belgium', '1000', '+32 02 219 03 03', None, 'daan_peeters@apple.be', 4),
 (9, 'Kara', 'Nielsen', None, 'Sønder Boulevard 51', 'Copenhagen', None, 'Denmark', '1720', '+453 3331 9991', None, 'kara.nielsen@jubii.dk', 4),
 (10, 'Eduardo', 'Martins', 'Woodstock Discos', 'Rua Dr. Falcão Filho, 155', 'São Paulo', 'SP', 'Brazil', '01007-010', '+55 (11) 3033-5446', '+55 (11) 3033-4564', 'eduardo@woodstock.com.br', 4)]

In [5]:
# practice - retrieve the "Genre" table
%sql SELECT * from "Genre" Limit 15

 * postgresql://postgres:***@localhost/Chinook
15 rows affected.


[(1, 'Rock'),
 (2, 'Jazz'),
 (3, 'Metal'),
 (4, 'Alternative & Punk'),
 (5, 'Rock And Roll'),
 (6, 'Blues'),
 (7, 'Latin'),
 (8, 'Reggae'),
 (9, 'Pop'),
 (10, 'Soundtrack'),
 (11, 'Bossa Nova'),
 (12, 'Easy Listening'),
 (13, 'Heavy Metal'),
 (14, 'R&B/Soul'),
 (15, 'Electronica/Dance')]

### sqlalchemy
The sqlalchemy library is another tool to establish a connection between PostgreSQL and Jupyter Notebook. This library can help you to perform better data analysis by, for example, combining the power of Pandas. First, you have to establish an engine to ensure a connection. Follow these prompts:

- Import create_engine from sqlalchemy.
- Specify a name for the new engine you are creating and link it to the connection_string you created with the ipython-sql library. This is another example of writing Pythonic code.
- Import the inspect package from sqlalchemy. This will help with querying the database. In this example, we want to determine the names of all the tables within the Chinook database.
- Hint: If you get an error message, simply rerun the code snippet. Python has to connect to PostgreSQL and then return the output. A small break in communication might result in an error message.

In [6]:
# Import necessary libraries to create an engine.
from sqlalchemy import create_engine

# Name the engine.
engine = create_engine(connection_string)

# Import inspect.
from sqlalchemy import inspect

insp = inspect(engine)
insp.get_table_names()

['Artist',
 'Album',
 'Employee',
 'Customer',
 'Invoice',
 'InvoiceLine',
 'Track',
 'Playlist',
 'PlaylistTrack',
 'Genre',
 'MediaType',
 'new_customer',
 'new_records',
 'mos_popular_artists',
 'peerreview']

### Confirm connection II
How and when does Pandas come into play? You can use Pandas to create a DataFrame directly from the Chinook database. Why is that important? If you have a Pandas DataFrame, you can tap into Python’s visualisation libraries and other data analysis tools. Follow these prompts for an example:

- Import the Pandas library because you are working in a new environment.
- Specify the name of the DataFrame you want to create.
- Indicate to Pandas where to get the raw data from with the pd.read_sql() function.
- Specify the SELECT statement, inside the brackets, to query the table directly from PostgreSQL. (Hint: The SQL syntax needs to be in quotation marks to be passed as a string.)
- Remember to specify the engine you created as a connector between PostgreSQL and Jupyter Notebook.
- Print the DataFrame.

In [7]:
# Import Pandas.
import pandas as pd

# Create a DataFrame.
df = pd.read_sql('SELECT * from "Artist" LIMIT 10', engine)

# View the DataFrame.
df

ArtistId                  Name
0         1                 AC/DC
1         2                Accept
2         3             Aerosmith
3         4     Alanis Morissette
4         5       Alice In Chains
5         6  Antônio Carlos Jobim
6         7          Apocalyptica
7         8            Audioslave
8         9              BackBeat
9        10          Billy Cobham

In [8]:
# Create a DataFrame.
customer_df = pd.read_sql('SELECT * from "Customer" LIMIT 10', engine)

# View the DataFrame.
customer_df

CustomerId  FirstName     LastName  \
0           1       Luís    Gonçalves   
1           2     Leonie       Köhler   
2           3   François     Tremblay   
3           4      Bjørn       Hansen   
4           5  František  Wichterlová   
5           6     Helena         Holý   
6           7     Astrid       Gruber   
7           8       Daan      Peeters   
8           9       Kara      Nielsen   
9          10    Eduardo      Martins   

                                            Company  \
0  Embraer - Empresa Brasileira de Aeronáutica S.A.   
1                                              None   
2                                              None   
3                                              None   
4                                  JetBrains s.r.o.   
5                                              None   
6                                              None   
7                                              None   
8                                              None   
9                                  Woodstock Discos   

                                Address                 City State  \
0       Av. Brigadeiro Faria Lima, 2170  São José dos Campos    SP   
1               Theodor-Heuss-Straße 34            Stuttgart  None   
2                     1498 rue Bélanger             Montréal    QC   
3                      Ullevålsveien 14                 Oslo  None   
4                         Klanova 9/506               Prague  None   
5                         Rilská 3174/6               Prague  None   
6  Rotenturmstraße 4, 1010 Innere Stadt               Vienne  None   
7                       Grétrystraat 63             Brussels  None   
8                   Sønder Boulevard 51           Copenhagen  None   
9             Rua Dr. Falcão Filho, 155            São Paulo    SP   

          Country PostalCode               Phone                 Fax  \
0          Brazil  12227-000  +55 (12) 3923-5555  +55 (12) 3923-5566   
1         Germany      70174    +49 0711 2842222                None   
2          Canada    H2G 1A7   +1 (514) 721-4711                None   
3          Norway       0171     +47 22 44 22 22                None   
4  Czech Republic      14700    +420 2 4172 5555    +420 2 4172 5555   
5  Czech Republic      14300    +420 2 4177 0449                None   
6         Austria       1010      +43 01 5134505                None   
7         Belgium       1000    +32 02 219 03 03                None   
8         Denmark       1720      +453 3331 9991                None   
9          Brazil  01007-010  +55 (11) 3033-5446  +55 (11) 3033-4564   

                      Email  SupportRepId  
0      luisg@embraer.com.br             3  
1     leonekohler@surfeu.de             5  
2       ftremblay@gmail.com             3  
3     bjorn.hansen@yahoo.no             4  
4  frantisekw@jetbrains.com             4  
5           hholy@gmail.com             5  
6    astrid.gruber@apple.at             5  
7     daan_peeters@apple.be             4  
8     kara.nielsen@jubii.dk             4  
9  eduardo@woodstock.com.br             4

In [9]:
# Create a DataFrame.
genre_df = pd.read_sql('SELECT * from "Genre" LIMIT 10', engine)

# View the DataFrame.
genre_df 

GenreId                Name
0        1                Rock
1        2                Jazz
2        3               Metal
3        4  Alternative & Punk
4        5       Rock And Roll
5        6               Blues
6        7               Latin
7        8              Reggae
8        9                 Pop
9       10          Soundtrack

In [10]:
# Export data as CSV.
genre_df.to_csv('genre_df.csv')
customer_df.to_csv('customer_df.csv')

## Apply Your Learning
Let’s use basic SQL syntax to query the Chinook database by utilising the connection you created earlier with the ipython-sql, sqlalchemy and psycopg2 libraries between PostgreSQL and Jupyter Notebook.

You'll construct three tables with basic SQL syntax in your Jupyter Notebook. The main aim of this exercise is to illustrate how to use SQL commands in combination with Python.

#### Construct a table indicating the first name, last name, and country of customers.
The three columns are all present in the Customer table. Therefore, you can use the SELECT statement to query the Customer table. The code snippet will be:

- Indicate to Python that you want to use SQL commands with %%sql.
- Specify the column names that you query (FirstName, LastName, and Country) with the SELECT statement.
- Use the FROM clause to specify the table as Customer.
- For illustrative purposes, LIMIT the query to 10 rows.

In [11]:
%%sql

SELECT "FirstName", "LastName", "Country"
FROM "Customer"
LIMIT 10

 * postgresql://postgres:***@localhost/Chinook
10 rows affected.


[('Luís', 'Gonçalves', 'Brazil'),
 ('Leonie', 'Köhler', 'Germany'),
 ('François', 'Tremblay', 'Canada'),
 ('Bjørn', 'Hansen', 'Norway'),
 ('Franti\x9aek', 'Wichterlová', 'Czech Republic'),
 ('Helena', 'Holý', 'Czech Republic'),
 ('Astrid', 'Gruber', 'Austria'),
 ('Daan', 'Peeters', 'Belgium'),
 ('Kara', 'Nielsen', 'Denmark'),
 ('Eduardo', 'Martins', 'Brazil')]

Look at that! You have a table with the FirstName, LastName and Country. There is one difference between the Jupyter Notebook code snippet and the pgAdmin SQL syntax. Can you spot the difference? Whenever you refer to a table in pgAdmin, you have to use public.table_name. Rerun the query with FROM public."Customer". What do you see?

Can you use the concat() function and combine the FirstName and LastName columns into a FullName column?

In [12]:
%%sql

SELECT CONCAT("FirstName", ' ', "LastName") AS "FullName", "Country"
FROM "Customer"
LIMIT 10

 * postgresql://postgres:***@localhost/Chinook
10 rows affected.


[('Luís Gonçalves', 'Brazil'),
 ('Leonie Köhler', 'Germany'),
 ('François Tremblay', 'Canada'),
 ('Bjørn Hansen', 'Norway'),
 ('Franti\x9aek Wichterlová', 'Czech Republic'),
 ('Helena Holý', 'Czech Republic'),
 ('Astrid Gruber', 'Austria'),
 ('Daan Peeters', 'Belgium'),
 ('Kara Nielsen', 'Denmark'),
 ('Eduardo Martins', 'Brazil')]

#### Determine the total of all the invoices generated during December 2009.
What if you want to determine the total amount of sales for December 2009? You will only need to query one table, Invoice, and the code snippet will be:

- Specify the SQL commands with %%sql.
- Specify that you want the SUM of the Total column with the SELECT statement.
- Use the FROM clause to specify the table as Invoice.
- Specify the dates with the BETWEEN and AND clauses.

In [13]:
%%sql

SELECT SUM("Total")
FROM "Invoice"
WHERE "InvoiceDate" BETWEEN '01/12/2009' AND '31/12/2009'

 * postgresql://postgres:***@localhost/Chinook
1 rows affected.


[(Decimal('37.62'),)]

Well done! The total amount of sales for December 2009 was $37.62. It doesn't seem high, but remember it was more than 12 years ago. Can you write a query to answer the following question?

Which year had the most sales in December?

In [14]:
%%sql

SELECT date_part('year', "InvoiceDate") AS year, date_part('month', "InvoiceDate") AS month, SUM("Total")
FROM "Invoice"
WHERE date_part('month', "InvoiceDate") = 12
GROUP BY year, month;

 * postgresql://postgres:***@localhost/Chinook
5 rows affected.


[(2009.0, 12.0, Decimal('37.62')),
 (2010.0, 12.0, Decimal('37.62')),
 (2011.0, 12.0, Decimal('37.62')),
 (2012.0, 12.0, Decimal('37.62')),
 (2013.0, 12.0, Decimal('38.62'))]

#### Construct a table to indicate the composer for each track and runtime in milliseconds.
There are a lot of tracks and each of them has a composer and runtime in milliseconds. The columns Composer, Name, and Milliseconds of the tracks are in the Track table. To construct a table indicating the composer and runtime of each track, do the following:

- Specify the SQL commands with %%sql.
- Specify that you want the columns Name, Composer, and Milliseconds with the SELECT statement.
- Use the FROM clause to specify the table as Track.
- Specify how the results table should be grouped with the GROUP BY function as Name, Composer and Milliseconds.
- The ORDER BY will indicate how the results table will be sorted.

In [15]:
%%sql

SELECT "Name", "Composer", "Milliseconds"
FROM "Track"
GROUP BY "Name", "Composer", "Milliseconds"
ORDER BY "Composer" DESC
LIMIT 10

 * postgresql://postgres:***@localhost/Chinook
10 rows affected.


[('The Real Problem', None, 11650),
 ('Geração Coca-Cola (Ao Vivo)', None, 228153),
 ('Se...', None, 286432),
 ('Um Certo Alguém', None, 194063),
 ('Nossa Gente (Avisa Là)', None, 188212),
 ('Samba Pra Endrigo', None, 259265),
 ('13 Years Of Grief', None, 246987),
 ('Branch Closing', None, 1814855),
 ('Oi, La', None, 167053),
 ('País Tropical', None, 452519)]

Congratulations! Using LIMIT the table was limited to 10 rows for illustrative purposes. Can you see that the composers are indicated as None? If you run the same query in pgAdmin, the absence of composers will be [null].

Can you determine the composer with the most tracks and longest runtime?

In [16]:
%%sql

SELECT "Composer", COUNT("Composer")
FROM "Track"
GROUP BY "Composer"
ORDER BY COUNT("Composer") DESC
LIMIT 10

 * postgresql://postgres:***@localhost/Chinook
10 rows affected.


[('Steve Harris', 80),
 ('U2', 44),
 ('Jagger/Richards', 35),
 ('Billy Corgan', 31),
 ('Kurt Cobain', 26),
 ('Bill Berry-Peter Buck-Mike Mills-Michael Stipe', 25),
 ('The Tea Party', 24),
 ('Chris Cornell', 23),
 ('Miles Davis', 23),
 ('Chico Science', 23)]

Track/Composer combo with longest run time

In [17]:
%%sql

SELECT "Name", "Composer", "Milliseconds"
FROM "Track"
GROUP BY "Name", "Composer", "Milliseconds"
ORDER BY "Milliseconds" DESC
LIMIT 10

 * postgresql://postgres:***@localhost/Chinook
10 rows affected.


[('Occupation / Precipice', None, 5286953),
 ('Through a Looking Glass', None, 5088838),
 ('Greetings from Earth, Pt. 1', None, 2960293),
 ('The Man With Nine Lives', None, 2956998),
 ('Battlestar Galactica, Pt. 2', None, 2956081),
 ('Battlestar Galactica, Pt. 1', None, 2952702),
 ('Murder On the Rising Star', None, 2935894),
 ('Battlestar Galactica, Pt. 3', None, 2927802),
 ('Take the Celestra', None, 2927677),
 ('Fire In Space', None, 2926593)]

Total runtime by Composer, all tracks

In [18]:
%%sql

SELECT "Composer", sum("Milliseconds") as "total_milliseconds"
FROM "Track"
GROUP BY "Composer"
ORDER BY sum("Milliseconds") DESC
LIMIT 10

 * postgresql://postgres:***@localhost/Chinook
10 rows affected.


[(None, 695840650),
 ('Steve Harris', 27217126),
 ('U2', 11271816),
 ('Miles Davis', 8996378),
 ('Jagger/Richards', 8667178),
 ('Billy Corgan', 7912553),
 ('The Tea Party', 6579763),
 ('Bill Berry-Peter Buck-Mike Mills-Michael Stipe', 6395130),
 ('Chris Robinson/Rich Robinson', 6059094),
 ('Chris Cornell', 5866463)]

#### Are you ready for a challenge? Would you be able to write a code snippet querying the Chinook database by yourself?
How many tracks have a runtime of fewer than 1,000,000 milliseconds?

In [19]:
%%sql

SELECT COUNT("Milliseconds")
FROM "Track"
WHERE "Milliseconds" < 1000000

 * postgresql://postgres:***@localhost/Chinook
1 rows affected.


[(3288,)]

## 5.3.4 Part II
Ryan, a data analyst with Know-your-Brand, is busy developing a machine learning (ML) model. The Chinook database is perfect to use as an example of a historical database. If the team can develop code that can provide answers to similar questions posed by Know-your-Brand, Ryan can develop the ML model based on the Chinook database. The ML model can then be tested on real data from Know-your-Brand. Therefore, Ryan asks you to answer the following questions:

- Who are the 15 most popular artists?
- What is the most sought genre based on media type?
- Which artist is the most popular based on genre?

Remember to break down a business question into smaller pieces. Just like a puzzle!

#### Question 1: Who are the 15 most popular artists?
This seems like a straightforward question. We need to rank the artist from most popular to least popular. However, there are no specifics on how to rank them. Do we have to rank the artists based on sales (invoices), number of tracks or number of albums?

The InvoiceLine table has a foreign key (FK) with the Track table, while the Track table is connected to the Album table with an FK. Lastly, the Album table is connected to the Artist table with an FK. The answer to this question does not seem that straightforward anymore. Let’s break it down into smaller parts.

The main question is to determine the 15 most popular artists. Therefore, we need the names of the artists. The Artist table only has two columns: ArtistId and Name. To determine the 15 most popular artists according to albums released, simply join the Artist and Album tables. The joined table will also give us the Title of the albums. However, it may be more accurate to determine the popularity of artists based on tracks released. The number of tracks on an album can differ. Therefore, we have to join the Track table as well. In essence, we will need to JOIN four different tables. 

The code snippet can be written as follows:

- Start with %%sql indicating to Jupyter Notebook that we are using SQL syntax.
- Specify the CREATE TABLE command to create a new table. This is only for future references and a good practice to get used to. This ensures that we don’t have to rewrite the query every time we want to return the specified table.
- Start the query with the SELECT statement indicating the specified column names we are interested in. In this example, we need the ArtistId, Name of the artist and album Title.
- Use the aggregate functions (COUNT and SUM) to filter the results. We are interested in the number of artists (SUM(ArtistId)).
- The FROM clause indicates the table we want to use as the starting point.
- Specify each of the tables we want to JOIN with the USING clause indicating the FK.
- Remember to use GROUP BY to indicate how the results table should be organised (column names to be displayed).
- Finally, the ORDER BY clause indicates how to sort the results table.
- Remember to use aliases to write more concise and precise syntax.

In [20]:
%%sql

CREATE TABLE most_popular_artists AS 
SELECT "ArtistId", "Artist"."Name", SUM("Quantity") AS "TracksSold"
FROM "Artist"
INNER JOIN "Album" USING ("ArtistId")
INNER JOIN "Track" USING ("AlbumId")
INNER JOIN "InvoiceLine" USING ("TrackId")
GROUP BY "ArtistId", "Artist"."Name"
ORDER BY SUM("Quantity") DESC

 * postgresql://postgres:***@localhost/Chinook
165 rows affected.


[]

This is the given answer i dont think its right as it breaks tracke by album and thats not the question is it plus also why would you sum Artist ID that just makes no sense


In [21]:

%%sql
SELECT ar."ArtistId", ar."Name", al."Title", 
COUNT(*),
SUM(ar."ArtistId")
FROM "Artist" ar 
JOIN "Album" al USING ("ArtistId")
JOIN "Track" tr USING ("AlbumId")
JOIN "InvoiceLine" i USING ("TrackId")
GROUP BY ar."ArtistId", ar."Name", al."Title"
ORDER BY COUNT(ar."ArtistId") DESC;

 * postgresql://postgres:***@localhost/Chinook
304 rows affected.


[(17, 'Chico Buarque', 'Minha Historia', 27, 459),
 (100, 'Lenny Kravitz', 'Greatest Hits', 26, 2600),
 (81, 'Eric Clapton', 'Unplugged', 25, 2025),
 (146, 'Titãs', 'Acústico', 22, 3212),
 (52, 'Kiss', 'Greatest Kiss', 20, 1040),
 (76, 'Creedence Clearwater Revival', 'Chronicle, Vol. 2', 19, 1444),
 (144, 'The Who', 'My Generation - The Very Best Of The Who', 19, 2736),
 (16, 'Caetano Veloso', 'Prenda Minha', 19, 304),
 (113, 'Os Paralamas Do Sucesso', 'Acústico MTV', 18, 2034),
 (158, 'Battlestar Galactica (Classic)', 'Battlestar Galactica (Classic), Season 1', 18, 2844),
 (76, 'Creedence Clearwater Revival', 'Chronicle, Vol. 1', 18, 1368),
 (54, 'Green Day', 'International Superhits', 18, 972),
 (150, 'U2', 'Rattle And Hum', 17, 2550),
 (88, "Guns N' Roses", 'Use Your Illusion I', 17, 1496),
 (69, 'Gene Krupa', "Up An' Atom", 17, 1173),
 (124, 'R.E.M.', 'The Best Of R.E.M.: The IRS Years', 17, 2108),
 (18, 'Chico Science & Nação Zumbi', 'Afrociberdelia', 16, 288),
 (156, 'The Office', 'The Office, Season 3', 16, 2496),
 (150, 'U2', 'Instant Karma: The Amnesty International Campaign to Save Darfur', 16, 2400),
 (132, 'Soundgarden', 'A-Sides', 15, 1980),
 (70, 'Toquinho & Vinícius', 'Vinícius De Moraes - Sem Limite', 15, 1050),
 (113, 'Os Paralamas Do Sucesso', 'Arquivo Os Paralamas Do Sucesso', 15, 1695),
 (81, 'Eric Clapton', 'The Cream Of Clapton', 15, 1215),
 (6, 'Antônio Carlos Jobim', 'Chill: Brazil (Disc 2)', 15, 90),
 (54, 'Green Day', 'American Idiot', 15, 810),
 (51, 'Queen', 'Greatest Hits II', 14, 714),
 (124, 'R.E.M.', 'New Adventures In Hi-Fi', 14, 1736),
 (148, 'Heroes', 'Heroes, Season 1', 13, 1924),
 (111, 'O Terço', 'Compositores', 13, 1443),
 (82, 'Faith No More', 'Album Of The Year', 13, 1066),
 (149, 'Lost', 'Lost, Season 2', 13, 1937),
 (82, 'Faith No More', 'Angel Dust', 13, 1066),
 (27, 'Gilberto Gil', 'Quanta Gente Veio Ver (Live)', 13, 351),
 (131, 'Smashing Pumpkins', 'Rotten Apples: Greatest Hits', 13, 1703),
 (41, 'Elis Regina', 'Elis Regina-Minha História', 13, 533),
 (252, 'Amy Winehouse', 'Back to Black', 12, 3024),
 (145, 'Tim Maia', 'Serie Sem Limite (Disc 1)', 12, 1740),
 (50, 'Metallica', 'ReLoad', 12, 600),
 (113, 'Os Paralamas Do Sucesso', 'Arquivo II', 12, 1356),
 (14, 'Bruce Dickinson', 'Chemical Wedding', 12, 168),
 (146, 'Titãs', 'Volume Dois', 12, 1752),
 (147, 'Battlestar Galactica', 'Battlestar Galactica, Season 3', 12, 1764),
 (135, 'System Of A Down', 'Mezmerize', 12, 1620),
 (50, 'Metallica', 'Garage Inc. (Disc 2)', 12, 600),
 (91, 'James Brown', 'Sex Machine', 12, 1092),
 (51, 'Queen', 'News Of The World', 12, 612),
 (77, 'Cássia Eller', 'Cássia Eller - Coleção Sem Limite [Disc 2]', 12, 924),
 (145, 'Tim Maia', 'Serie Sem Limite (Disc 2)', 12, 1740),
 (150, 'U2', 'The Best Of 1980-1990', 11, 1650),
 (150, 'U2', 'B-Sides 1980-1990', 11, 1650),
 (131, 'Smashing Pumpkins', 'Judas 0: B-Sides and Rarities', 11, 1441),
 (53, 'Spyro Gyra', 'Heart of the Night', 11, 583),
 (50, 'Metallica', "Kill 'Em All", 11, 550),
 (50, 'Metallica', 'Load', 11, 550),
 (51, 'Queen', 'Greatest Hits I', 11, 561),
 (52, 'Kiss', 'Unplugged [Live]', 11, 572),
 (90, 'Iron Maiden', 'The X Factor', 11, 990),
 (13, 'Body Count', 'Body Count', 11, 143),
 (21, 'Various Artists', 'Vozes do MPB', 11, 231),
 (150, 'U2', 'War', 11, 1650),
 (149, 'Lost', 'Lost, Season 3', 11, 1639),
 (152, 'Van Halen', 'The Best Of Van Halen, Vol. I', 10, 1520),
 (126, 'Raul Seixas', 'Raul Seixas', 10, 1260),
 (3, 'Aerosmith', 'Big Ones', 10, 30),
 (50, 'Metallica', 'Master Of Puppets', 10, 500),
 (88, "Guns N' Roses", 'Use Your Illusion II', 10, 880),
 (1, 'AC/DC', 'For Those About To Rock We Salute You', 10, 10),
 (134, 'Stone Temple Pilots', 'Core', 10, 1340),
 (22, 'Led Zeppelin', 'BBC Sessions [Disc 2] [Live]', 10, 220),
 (98, 'Judas Priest', 'Living After Midnight', 10, 980),
 (143, 'The Tea Party', 'Tangents', 10, 1430),
 (114, 'Ozzy Osbourne', 'Tribute', 10, 1140),
 (112, 'Olodum', 'Olodum', 10, 1120),
 (127, 'Red Hot Chili Pe

In [22]:
# Import the SQL query to pandas
import pandas as pd

# Create a DataFrame
q1 = pd.read_sql('SELECT * FROM most_popular_artists', engine)

# View the output.
q1

ArtistId                                               Name  TracksSold
0          90                                        Iron Maiden         140
1         150                                                 U2         107
2          50                                          Metallica          91
3          22                                       Led Zeppelin          87
4         113                            Os Paralamas Do Sucesso          45
..        ...                                                ...         ...
160       214  Academy of St. Martin in the Fields & Sir Nevi...           1
161       157                                     Dread Zeppelin           1
162       222  Academy of St. Martin in the Fields, John Birc...           1
163       247                                 The King's Singers           1
164       260                     Adrian Leaper & Doreen de Feis           1

[165 rows x 3 columns]

In [23]:
# Export data as CSV.
q1.to_csv('q1.csv')

#### Question 2: What is the most sought genre based on the media type?
This question asks us to determine the most popular genre based on the media type. From the start, it looks like we will only need two tables: Genre and MediaType. Unfortunately, there is no key between these two tables. The only table that has both genre and media type as column names is the Track table. Seems like we have to JOIN three tables!

The code snippet can be written as follows:

- Start with %%sql indicating to Jupyter Notebook that we are using SQL syntax.
- Specify the CREATE TABLE command to create a new table for future reference.
- Start the query with the SELECT statement indicating the specified column names we are interested in. In this example, we need the GenreId, Name of the Genre and MediaType Name. There are two columns with the name 'Name'. Can we still remember how to distinguish between them? (Hint: Use aliases.)
- Use the aggregate functions (COUNT and SUM) to filter the results. 
- The FROM clause indicates the table we want to use as the starting point.
- Specify the tables we want to JOIN with the USING clause indicating the key(s).
- Remember to use GROUP BY to indicate how the results table should be organised (column names to be displayed).
- Employ aliases to write more concise and precise syntax.
- Finally, the ORDER BY clause indicates how to sort the results table.

In [24]:
%%sql

CREATE TABLE genre_media AS
SELECT COUNT(g."GenreId"), m."MediaTypeId", m."Name" AS "MediaName", g."Name" AS "GenreName"
FROM "Genre" g
INNER JOIN "Track" t USING ("GenreId")
INNER JOIN "MediaType" m USING ("MediaTypeId")
GROUP BY m. "MediaTypeId", m."Name", g."Name"
ORDER BY COUNT("GenreId") DESC

 * postgresql://postgres:***@localhost/Chinook
38 rows affected.


[]

In [25]:
# Create a DataFrame.
q2 = pd.read_sql('SELECT * FROM genre_media', engine)

# View the output.
q2

count  MediaTypeId                    MediaName           GenreName
0    1211            1              MPEG audio file                Rock
1     578            1              MPEG audio file               Latin
2     374            1              MPEG audio file               Metal
3     332            1              MPEG audio file  Alternative & Punk
4     127            1              MPEG audio file                Jazz
5      93            3  Protected MPEG-4 video file            TV Shows
6      84            2     Protected AAC audio file                Rock
7      81            1              MPEG audio file               Blues
8      67            2     Protected AAC audio file           Classical
9      64            3  Protected MPEG-4 video file               Drama
10     58            1              MPEG audio file              Reggae
11     49            1              MPEG audio file            R&B/Soul
12     42            1              MPEG audio file          Soundtrack
13     38            2     Protected AAC audio file         Alternative
14     35            1              MPEG audio file         Hip Hop/Rap
15     34            2     Protected AAC audio file                 Pop
16     28            1              MPEG audio file   Electronica/Dance
17     28            1              MPEG audio file         Heavy Metal
18     26            1              MPEG audio file               World
19     26            3  Protected MPEG-4 video file    Sci Fi & Fantasy
20     24            1              MPEG audio file      Easy Listening
21     17            3  Protected MPEG-4 video file              Comedy
22     15            1              MPEG audio file          Bossa Nova
23     14            1              MPEG audio file                 Pop
24     13            3  Protected MPEG-4 video file     Science Fiction
25     12            1              MPEG audio file       Rock And Roll
26     12            2     Protected AAC audio file            R&B/Soul
27      6            4     Purchased AAC audio file           Classical
28      3            5               AAC audio file                Jazz
29      2            5               AAC audio file   Electronica/Dance
30      2            5               AAC audio file                Rock
31      2            5               AAC audio file               World
32      1            3  Protected MPEG-4 video file         Alternative
33      1            4     Purchased AAC audio file         Alternative
34      1            2     Protected AAC audio file          Soundtrack
35      1            5               AAC audio file           Classical
36      1            2     Protected AAC audio file               Opera
37      1            5               AAC audio file               Latin

#### Question 3: Which artist is the most popular based on genre?
Now, we need to determine the most popular artist based on genre. From the start, it looks like we will only need two tables: Genre and Artist. Again, there is no key between these two tables. The Artist and Album tables are connected with a key, while the Genre, Track and Album tables are connected with keys. Seems like we have to JOIN four tables!

The code snippet can be written similar to Questions 1 and 2:

- Start with %%sql indicating to Jupyter Notebook that we are using SQL syntax.
- Specify the CREATE TABLE command to create a new table for future reference.
- Start the query with the SELECT statement indicating the specified column. 
- Use the aggregate functions (COUNT and SUM) to filter the results. 
- The FROM clause indicates the table we want to use as the starting point.
- Specify the tables we want to JOIN with the USING clause indicating the key(s).
- Remember to use aliases to write more concise and precise syntax.
- Finally, the GROUP BY and ORDER BY clauses indicate how to sort the results table.

In [26]:
%%sql

CREATE TABLE artist_genre AS
SELECT g."Name" AS genre, ar."Name" AS artist_name, COUNT("Quantity") AS tracks_sold
FROM "Genre" g
INNER JOIN "Track" t USING ("GenreId")
INNER JOIN "InvoiceLine" i USING ("TrackId")
INNER JOIN "Album" al USING ("AlbumId")
INNER JOIN "Artist" ar USING ("ArtistId")
GROUP BY g."Name", ar."Name"
ORDER BY COUNT("Quantity") DESC

 * postgresql://postgres:***@localhost/Chinook
194 rows affected.


[]

In [27]:
# Create a DataFrame.
q3 = pd.read_sql('SELECT * FROM artist_genre', engine)

# View the output.
q3

genre                                        artist_name  tracks_sold
0        Metal                                          Metallica           91
1         Rock                                                 U2           91
2         Rock                                       Led Zeppelin           87
3        Metal                                        Iron Maiden           70
4         Rock                                        Iron Maiden           54
..         ...                                                ...          ...
189  Classical              Orchestra of The Age of Enlightenment            1
190  Classical  Academy of St. Martin in the Fields, John Birc...            1
191  Classical           Antal Doráti & London Symphony Orchestra            1
192  Classical             Berliner Philharmoniker & Hans Rosbaud            1
193  Classical                                       Julian Bream            1

[194 rows x 3 columns]

#### Apply your learning
Are you ready for a challenge? See if you can answer this problem on your own.

Determine the most popular artist based on genre, tracks sold and customer country.

In [28]:
%%sql

SELECT g."Name" AS genre, ar."Name" AS artist_name, "Country", COUNT("Quantity") AS tracks_sold
FROM "Genre" g
INNER JOIN "Track" t USING ("GenreId")
INNER JOIN "InvoiceLine" i USING ("TrackId")
INNER JOIN "Album" al USING ("AlbumId")
INNER JOIN "Artist" ar USING ("ArtistId")
INNER JOIN "Invoice" i2 USING ("InvoiceId")
INNER JOIN "Customer" c USING ("CustomerId")
GROUP BY g."Name", ar."Name", "Country"
ORDER BY COUNT("Quantity") DESC

 * postgresql://postgres:***@localhost/Chinook
801 rows affected.


[('Metal', 'Metallica', 'USA', 27),
 ('Rock', 'U2', 'USA', 23),
 ('Latin', 'Os Paralamas Do Sucesso', 'Canada', 16),
 ('Metal', 'Metallica', 'Canada', 15),
 ('Metal', 'Iron Maiden', 'USA', 15),
 ('Rock', 'Deep Purple', 'USA', 15),
 ('Rock', 'Led Zeppelin', 'Canada', 14),
 ('Rock', 'Iron Maiden', 'USA', 14),
 ('Rock', 'Led Zeppelin', 'USA', 13),
 ('Rock', 'Led Zeppelin', 'Germany', 12),
 ('Latin', 'Tim Maia', 'USA', 12),
 ('Latin', 'Caetano Veloso', 'USA', 12),
 ('Latin', 'Os Paralamas Do Sucesso', 'Brazil', 11),
 ('Rock', 'Creedence Clearwater Revival', 'France', 11),
 ('Alternative & Punk', 'Green Day', 'USA', 11),
 ('Rock', 'Pearl Jam', 'Brazil', 10),
 ('Blues', 'Eric Clapton', 'USA', 10),
 ('Rock', 'Led Zeppelin', 'France', 10),
 ('Latin', 'Antônio Carlos Jobim', 'USA', 10),
 ('Metal', 'Metallica', 'France', 10),
 ('Latin', 'Chico Science & Nação Zumbi', 'Brazil', 9),
 ('Alternative & Punk', 'R.E.M.', 'Canada', 9),
 ('Rock', 'The Cult', 'USA', 9),
 ('Rock', 'Led Zeppelin', 'Norway', 9),
 ('Rock', 'Iron Maiden', 'Australia', 9),
 ('TV Shows', 'The Office', 'USA', 9),
 ('Rock', 'Queen', 'Canada', 9),
 ('Rock', 'U2', 'Portugal', 9),
 ('Metal', 'Metallica', 'United Kingdom', 9),
 ('Rock', 'Kiss', 'Canada', 9),
 ('Rock', 'U2', 'Austria', 9),
 ('Metal', 'Metallica', 'Germany', 9),
 ('R&B/Soul', 'Amy Winehouse', 'USA', 9),
 ('Rock', 'Deep Purple', 'Germany', 9),
 ('Rock', "Guns N' Roses", 'Brazil', 9),
 ('Alternative & Punk', 'Faith No More', 'USA', 8),
 ('Rock', 'The Rolling Stones', 'USA', 8),
 ('Rock', 'U2', 'Canada', 8),
 ('Metal', 'Iron Maiden', 'Portugal', 8),
 ('Rock', "Guns N' Roses", 'USA', 8),
 ('Rock', 'Van Halen', 'Germany', 8),
 ('Comedy', 'The Office', 'USA', 8),
 ('Rock', 'Van Halen', 'USA', 8),
 ('Rock', 'Deep Purple', 'Brazil', 7),
 ('Alternative & Punk', 'Faith No More', 'Belgium', 7),
 ('Bossa Nova', 'Toquinho & Vinícius', 'Canada', 7),
 ('Latin', 'Chico Buarque', 'France', 7),
 ('Metal', 'Iron Maiden', 'United Kingdom', 7),
 ('Rock', 'Creedence Clearwater Revival', 'Denmark', 7),
 ('Rock', 'Soundgarden', 'USA', 7),
 ('Rock', 'Led Zeppelin', 'United Kingdom', 7),
 ('Rock', 'Deep Purple', 'India', 7),
 ('Bossa Nova', 'Toquinho & Vinícius', 'USA', 7),
 ('Latin', 'Lulu Santos', 'United Kingdom', 7),
 ('Rock', 'U2', 'Brazil', 7),
 ('TV Shows', 'Lost', 'Ireland', 7),
 ('Blues', 'Eric Clapton', 'Germany', 6),
 ('Alternative & Punk', 'Titãs', 'USA', 6),
 ('Rock', 'Pearl Jam', 'Canada', 6),
 ('Latin', 'Chico Science & Nação Zumbi', 'Canada', 6),
 ('Rock', 'Led Zeppelin', 'Chile', 6),
 ('Latin', 'Tim Maia', 'Brazil', 6),
 ('Metal', 'Iron Maiden', 'Germany', 6),
 ('Rock', 'U2', 'Ireland', 6),
 ('Latin', 'Djavan', 'Germany', 6),
 ('Latin', 'Various Artists', 'USA', 6),
 ('Alternative & Punk', 'Faith No More', 'Germany', 6),
 ('Metal', 'Iron Maiden', 'India', 6),
 ('Rock', 'Van Halen', 'Finland', 6),
 ('Metal', 'Judas Priest', 'Canada', 6),
 ('Metal', 'Iron Maiden', 'Canada', 6),
 ('Latin', 'Chico Science & Nação Zumbi', 'Ireland', 6),
 ('Rock', 'U2', 'Poland', 6),
 ('Rock', 'Foo Fighters', 'USA', 6),
 ('Alternative & Punk', 'Body Count', 'France', 6),
 ('Jazz', 'Spyro Gyra', 'USA', 6),
 ('Metal', 'Metallica', 'Argentina', 6),
 ('Jazz', 'Gene Krupa', 'India', 6),
 ('Rock', 'Men At Work', 'Brazil', 6),
 ('Metal', 'Metallica', 'Brazil', 6),
 ('TV Shows', 'Lost', 'Czech Republic', 5),
 ('Latin', 'Milton Nascimento', 'Canada', 5),
 ('Rock', 'Pearl Jam', 'Spain', 5),
 ('Rock', 'Queen', 'France', 5),
 ('Alternative & Punk', 'Smashing Pumpkins', 'Canada', 5),
 ('Latin', 'Chico Buarque', 'Brazil', 5),
 ('Metal', 'Iron Maiden', 'Brazil', 5),
 ('Rock', 'The Rolling Stones', 'Belgium', 5),
 ('Rock', 'Iron Maiden', 'Portugal', 5),
 ('Latin', 'Chico Buarque', 'USA', 5),
 ('Alternative & Punk', 'R.E.M.', 'USA', 5),
 ('Latin', 'Various Artists', 'France', 5),
 ('Rock', 'Iron Maiden', 'Czech Republic', 5),
 ('Latin', 'Cássia Eller', 'Poland', 5),
 ('Metal', 'Iron Maiden', 'Australia', 5),
 ('Rock', 'Queen', 'USA', 5),
 ('Latin', 'Milton Na